grails-app/services/streama/OpensubtitlesService.groovy

```diff
@@ -0,0 +1,168 @@
+      response.body.findAll { it.subFormat.equals("srt") || it.subFormat.equals("vtt") }.sort { -it.subDownloadsCnt }
+    }
+    return response
+  }
+
+  def getSubtitlesByHash(def videoId) {
+    def video = Video.findById(videoId)
+    Set subtitleUrlList = []
+    for (file in video.files) {
+      if (fileService.allowedVideoFormats.contains(file.extension)) {
+        if (file.openSubtitleHash?.isEmpty() ?: true) {
+          def filePath = uploadService.getPath(file)
+          try {
+            file.openSubtitleHash = openSubtitlesHasherService.computeHash(new java.io.File(filePath))
+            file.save flush: true
+          } catch (Exception e) {
+            log.error("Failed to save openSubtitleHash, ${e.message}", e)
+          }
+        }
+        subtitleUrlList.add(buildUrl(file.openSubtitleHash, file.size))
+      }
+    }
+    Set responseSet = []
+    def credentials = settingsService.getValueForName('credentials_opensubtitles')
+
+    for (url in subtitleUrlList) {
+      def response = sendRequest(url, credentials)
+      if (response.statusCodeValue != 200) {
+        return response
+      }
+      responseSet.addAll(response.body)
+    }
+    responseSet?.sort { -it.subDownloadsCnt }
+    def response = ResponseEntity.status(200).body(responseSet)
+    return response
+  }
+
+  def downloadSubtitles(String subtitleName, String url, String subtitleLanguage, videoId) {
+    def videoInstance = Video.findById(videoId)
+    if (videoInstance != null) {
+      def stagingDir = uploadService.getDir().uploadDir.toString()
+      def filename = url.tokenize('/')[-1]
+      def filePath = new java.io.File("$stagingDir/$filename")
+
+      new FileOutputStream(filePath).withStream { out ->
+        new URL(url).openStream().eachByte {
+          out.write(it)
+        }
+      }
+
+      try {
+        def fileName = zipService.unzipFile(stagingDir, filename, subtitleName)
+        def fileDb = saveFile(stagingDir, fileName, subtitleName, subtitleLanguage)
+        if (fileDb != null) {
+          if (videoInstance.getSubtitles().isEmpty()) {
+            fileDb.setIsDefault(true)
+          }
+          videoInstance.addToFiles(fileDb)
+          videoInstance.save flush: true, failOnError: true
+        }
+        return true
+      } catch (Exception e) {
+        log.error(e.message, e)
+        return false
+      } finally {
+        filePath.delete()
+      }
+    }
+  }
+
+  def sendRequest(String url, String credentials) {
+    HttpHeaders headers = new HttpHeaders()
+    def encoderCredentials = credentials.encodeAsBase64()
+    headers.set("Authorization", "Basic $encoderCredentials")
+    headers.set("User-Agent", "curl/7.65.3")
+    HttpEntity<String> entity = new HttpEntity<String>(headers)
+    def response = ResponseEntity.status(400).body("Opensubtitle API problems")
+    try {
+      response = restTemplate.exchange(url, HttpMethod.GET, entity, SubtitlesResponse[].class)
```

Comment:
Ive never used restTemplate, so this basically does a fancy casting of the response as a List of SubtitlesResponses?